### PR TITLE
Feature/sql permission checks

### PR DESCRIPTION
**Pull request by ulferts**
Reimplements permission checks in Arel/SQL which in turn improves extensibility and performance. Adds various performance improvements on other layers.

### CODE REVIEW

File: app/models/authorization.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  end
+
+  def self.add_scope_method(name, allowance)
+    method_body = ->(options = {}) { allowance.scope(options) }
+
+    eigenclass.send(:define_method, name, method_body)
+  end
+
+  def self.eigenclass
+    class << self; self; end;
```
Comment (myabc) [inline app/models/authorization.rb]: @ulferts can you not use `Kernel.singleton_class`?

File: app/models/authorization/condition/concatenation.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      self.second = second
+
+      super(scope)
+    end
+
+    def arel_statement(options)
+      if apply_condition?(options)
+        concat([first, second], options)
+      else
+        nil
```
Comment (myabc) [inline app/models/authorization/condition/concatenation.rb:45]: @ulferts any reason for the explicit `nil`?

File: spec/views/api/experimental/projects/show_api_json_spec.rb [hunk 3]
```diff
@@ -33,11 +33,13 @@
...
-    possible_responsible_members: members,
-    possible_assignee_members:    members,
-    types:                        types
-  )}
+  let(:project) do
+    p = FactoryGirl.build_stubbed(:project, types: types)
+    allow(p).to receive(:possible_responsibles).and_return(members)
+    allow(p).to receive(:possible_assignees).and_return(members)
+
+    p
```
Comment (myabc) [inline spec/views/api/experimental/projects/show_api_json_spec.rb]: see https://github.com/opf/openproject/pull/1760#discussion_r16345479

**Comment by myabc** [comment]
I made a few Ruby language-related comments. I don't understand the current permissions system fully, so I don't feel sufficiently qualified to give this PR the review it deserves.

**Comment by tessi** [comment]
Density (=number of errors found) vs. lines of code in a code review:

source: http://www.ibm.com/developerworks/rational/library/11-proven-practices-for-peer-review/

Having said that, I'll have a look at this PR :8ball:

File: app/controllers/api/v2/projects_controller.rb [hunk 4]
```diff
@@ -49,7 +49,7 @@ def index
           ids = ids.map(&:to_i).sort
           identifiers = identifiers.sort
 
-          options[:conditions] = ["id IN (?) OR identifier IN (?)", ids, identifiers]
+          options[:conditions] = ["projects.id IN (?) OR projects.identifier IN (?)", ids, identifiers]
```
Comment (tessi) [inline app/controllers/api/v2/projects_controller.rb]: Shouldn't we use `Project.table_name` here?

File: lib/open_project/needs_authorization/deprecated_methods.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  module DeprecatedMethods
+    # TODO: remove once the lazy User.current parameter is removed from method calls
+    def visible?(user = nil)
+      if user.nil?
+        warn <<-MESSAGE
+
+[DEPRECATION] Calling #{self.class}.visible? without passing a user is deprecated
+
+Called from #{caller[0]}
+        MESSAGE
```
Comment (tessi) [inline lib/open_project/needs_authorization/deprecated_methods.rb:40]: :thumbsup: for deprecating this. But deprecations should use the Deprecation logger :)

`ActiveSupport::Deprecation.warn <<-MESSAGE ...`

**Comment by tessi** [comment]
FYI: I resolved the merge conflicts and did a force-push

**Comment by floriank** [comment]
I have been lead here by https://github.com/finnlabs/openproject-impermanent_memberships/pull/41 - any update on this?

**Pull request closed without merging**